### PR TITLE
feat(linter): add `import/no-commonjs` rule

### PR DESCRIPTION
**Pull request by pumano**

### CODE REVIEW

**Comment by pumano** [comment]
This is my first time I trying to port rule, I tried my best, but few cases I can't cover (working with scopes). I added `FIXME:` labels, where problems exist.

File: crates/oxc_linter/src/rules/import/no_commonjs.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{GetSpan, Span};
+
+use oxc_ast::{ast::Expression, AstKind};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+fn no_commonjs_diagnostic(span: Span, name: &str, actual: &str) -> OxcDiagnostic {
+    // See <https://oxc.rs/docs/contribute/linter/adding-rules.html#diagnostics> for details
+    OxcDiagnostic::warn("Forbid CommonJS `require` calls and `module.exports` or `exports.*")
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/import/no_commonjs.rs:11]: Please follow guidance on crafting diagnostics messages in the above link.

Comment (DonIsaac) [reply crates/oxc_linter/src/rules/import/no_commonjs.rs:11]: Your help message should be the diagnostic message.

File: crates/oxc_linter/src/rules/import/no_commonjs.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    #[allow(clippy::redundant_else)]
+    if is_cond {
+        #[allow(clippy::needless_return)]
+        return true;
+    } else {
+        let Some(parent) = ctx.nodes().parent_node(parent_node.id()) else {
+            return false;
+        };
+        #[allow(clippy::needless_return)]
+        return is_conditional(parent, ctx);
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/import/no_commonjs.rs:127]:
```suggestion

       is_conditional(parent, ctx)
```

File: crates/oxc_linter/src/rules/import/no_commonjs.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+                };
+
+                if !member_expr.object().is_specific_id("module")
+                    && !member_expr.object().is_specific_id("exports")
+                    && property_name != "exports"
+                {
+                    return;
+                }
+
+                if member_expr.object().is_specific_id("module") && property_name == "exports" {
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/import/no_commonjs.rs:165]: This if statement does redundant checks, see if you can refactor these two into a single `if` or `match`

File: crates/oxc_linter/src/rules/import/no_commonjs.rs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+                    // FIXME: add proper ctx.nodes().iter_parents instead of 3 parent nodes below
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/import/no_commonjs.rs:166]: consider `iter_parents().next()` and then call `iter.next()` three times.

**Review by DonIsaac — COMMENTED** [review]
Looks like there's still a bunch of FIXME comments that need resolving.

**Comment by DonIsaac** [comment]
Marking this PR as a draft until FIXMEs are resolved.

File: crates/oxc_linter/src/rules/import/no_commonjs.rs (hunk 3 above)
Comment (pumano) [reply crates/oxc_linter/src/rules/import/no_commonjs.rs:156]: I remove previous check and that check is fit better as original rule.

**Comment by pumano** [comment]
@DonIsaac Looks like I can fix all issues. Please review

**Review by DonIsaac — APPROVED** [review]

**Pull request merged**